source/gui/addonStoreGui/viewModels/store.py

```diff
@@ -612,6 +619,46 @@ def _getAvailableAddonsInBG(self):
 		core.callLater(delay=0, callable=self.detailsVM.updated.notify, addonDetailsVM=self.detailsVM)
 		log.debug("completed refresh")
 
+	def _cancelDownloadForAddon(self, listItemVM: AddonListItemVM[_AddonStoreModel]):
+		try:
+			addonDataManager._downloadsPendingCompletion.remove(listItemVM)
+		except KeyError:
+			log.debug("Download already completed")
+		else:
+			futuresCopy = self._downloader._pending.copy()
+			for future, addon in futuresCopy.items():
+				if listItemVM == addon[0] and not future.cancel():
+					self._downloader.progress.pop(addon[0], None)
+		finally:
+			listItemVM.status = getStatus(listItemVM.model, self._filteredStatusKey)
+
+	def _cancelPendingInstallForAddon(self, listItemVM: AddonListItemVM[_AddonStoreModel]):
+		pendingInstallCopy = addonDataManager._downloadsPendingInstall.copy()
+		for addonData, fileDownloaded in pendingInstallCopy:
+			if addonData == listItemVM:
+				addonDataManager._downloadsPendingInstall.remove((addonData, fileDownloaded))
+				listItemVM.status = getStatus(listItemVM.model, self._filteredStatusKey)
+				# Clean up download file
+				try:
+					os.remove(fileDownloaded)
+				except FileNotFoundError:
+					pass  # File not found is benign and expected in some cases
+				except Exception as e:
+					log.error(f"Failed to delete downloaded file {fileDownloaded}: {e}")
+
+	def cancelInstallForAddon(self, listItemVM: AddonListItemVM[_AddonStoreModel]):
+		log.debug(f"Cancelling install of {listItemVM.Id}")
+
+		if listItemVM.status == AvailableAddonStatus.DOWNLOADING:
+			self._cancelDownloadForAddon(listItemVM)
+		elif listItemVM.status == AvailableAddonStatus.DOWNLOAD_SUCCESS:
+			self._cancelPendingInstallForAddon(listItemVM)

```

Comment:
```suggestion
			self._cancelPendingInstallForAddon(listItemVM)
		listItemVM.status = getStatus(listItemVM.model, self._filteredStatusKey)
```

Reply:
@seanbudd 

I was thinking that cleaning up the cancellation of an add-on that is downloading relies on a change in the state of the add-on, so is it possible that after `_downloadsPendingCompletion.remove(listItemVM)` the download is just about complete and the state is still DOWNLOADING?

Is there a way to pause the downloader thread from running while the installation is canceled?

Reply:
I don't think so, this approach needs to be thread safe. you could try adding a threading lock so that a download doesn't occur during the cancellation

Reply:
I'm sorry, but I don't know much about multithreading. Any hints?

Reply:
@coderabbitai can you explain how to use locks to make cancelInstallForAddon thread safe with AddonFileDownloader._executor tasks